File: graphr/src/graphr/graph/Vertex.java
package graphr.graph;

//import graphr.data.JsonArrayState;
//import graphr.data.JsonKeyValueState;
//import graphr.data.JsonReadableWritable;

import graphr.graph.Edge.Direction;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Hashtable;
import java.util.List;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class Vertex<DV extends GraphData, DE extends GraphData> extends
		GraphElement<DV> {
	private static Logger log = LogManager.getLogger();

	Hashtable<Long, Edge<DV, DE>> edges;

	public Vertex(DV data) {
		super(data);
		edges = new Hashtable<Long, Edge<DV, DE>>();
	}

	public void addEdge(Edge<DV, DE> e) {
		edges.put(new Long(e.id), e);
	}

	/**
	 * Returns outgoing edges.
	 * 
	 * @return
	 */
	public Collection<Edge<DV, DE>> getEdges() {
		return getEdges(Direction.OUTGOING);
	}

	/**
	 * Returns incoming and/or outgoing edges.
	 * 
	 * @param direction
	 * @return
	 */
	public Collection<Edge<DV, DE>> getEdges(Direction direction) {
		Collection<Edge<DV, DE>> coll = edges.values();

		if (direction.equals(Direction.BOTH))
			return coll;

		List<Edge<DV, DE>> outgoingEdges = new ArrayList<Edge<DV, DE>>();

		for (Edge<DV, DE> e : coll) {
			if (e.getSource().equals(this)) {
				outgoingEdges.add(e);
			}
		}

		if (direction.equals(Direction.OUTGOING)) {
			return outgoingEdges;
		} else if (direction.equals(Direction.INCOMING)) {
			coll.removeAll(outgoingEdges);
			return coll;
		}

		log.error("Unknown Direction " + direction);
		return null;
	}

	public Edge<DV, DE> getEdge(long id) {
		return edges.get(id);
	}

	public String toString() {
		return "Vertex (id=" + id + ")";
	}

	/**
	 * Part of the visitor design pattern -accept method.
	 */
	@Override
	public void accept(GraphElementVisitor visitor) {
		visitor.visit(this);
	}

	// public String getAsJson() {
	//
	// JsonKeyValueState j = new JsonKeyValueState();
	//
	// j.add("type", "Vertex");
	// j.add("id", new Integer(id).toString());
	// j.add("data", data != null ? data.getAsJson() : "null");
	//
	// JsonArrayState edgesForJson = new JsonArrayState();
	//
	// for (Edge<DV,DE> e : edges.values()) {
	// edgesForJson.add(e.getAsJson());
	// }
	//
	// j.add("edges", edgesForJson.getAsJson());
	//
	// return j.getAsJson();
	// }

}

File: graphrs/graphrs/graph/Vertex.swift
//
//  Vertex.swift
//  graphrs
//
//  Created by Peter Janacik on 23.12.15.
//  Copyright © 2015 CITBDA. All rights reserved.
//

public class Vertex: GraphElement, Hashable {
    private var edges = [UInt64: Edge]()
    
    public func addEdge(e: Edge) {
        edges[e.id] = e
    }
    
    func removeEdgeOnBothSides(e: Edge) {
        e.sideA.edges.removeValueForKey(e.id)
        e.sideB.edges.removeValueForKey(e.id)
    }
    
    /**
     * Returns outgoing edges.
     *
     * @return
     */
    public func getEdges() -> [Edge]? {
        return getEdges(Direction.Forward)
    }
    
    /**
     * Returns incoming and/or outgoing edges.
     *
     * @param direction
     * @return
     */
    public func getEdges(direction: Direction) -> [Edge]? {
        let edgeCollection = Array(edges.values)
        
        if (direction == Direction.Undefined) {
            return edgeCollection
        }
        
        var incommingEdges = [Edge]()
        var outgoingEdges = [Edge]()
        var bidirectEdges = [Edge]()
        
        for e in edgeCollection {
            switch e.direction {
                case .Forward:
                    outgoingEdges.append(e)
                    break
                
                case .Backward:
                    incommingEdges.append(e)
                    break
                    
                case .Bi:
                    bidirectEdges.append(e)
                    break
                    
                default:
                    break
            }
        }
        
        switch direction {
            case .Forward:
                return outgoingEdges
                
            case .Backward:
                return incommingEdges
                
            case .Bi:
                return bidirectEdges
                
            default:
                break
        }
        
        print("Unknown Direction \(direction)")
        return nil
    }
    
    public func getEdge(id: UInt64) -> Edge? {
        return edges[id]
    }
    
    public func getEdgeCount() -> Int {
        return edges.count
    }
    
    public func hasOutgoingEdge() -> Bool {
        return numberOutgoingEdges() > 0
    }
    
    public func hasIncommingEdge() -> Bool {
        return numberIncomingEdges() > 0
    }
    
    public func numberOutgoingEdges() -> Int {
        return getOutgoingEdges().count
    }
    
    public func numberIncomingEdges() -> Int {
        return getIncomingEdges().count
    }
    
    public func getOutgoingEdges() -> [Edge] {
        return getEdges(.Forward) ?? [Edge]()
    }
    
    public func getIncomingEdges() -> [Edge] {
        return getEdges(.Backward) ?? [Edge]()
    }
    
    public func toString() -> String {
        return "Vertex (id=\(id))"
    }
    
    /**
     * Part of the visitor design pattern -accept method.
     */
    override func accept(visitor: GraphElementVisitor) {
        visitor.visit(self)
    }
    
    public var hashValue: Int {
        return self.toString().hashValue
    }
}

public func ==(lhs: Vertex, rhs: Vertex) -> Bool {
    if (lhs.id == rhs.id && (lhs.data as! DictionaryElementData) == (rhs.data as! DictionaryElementData)) {
        return true
    }
    
    return false
}

public func !=(lhs: Vertex, rhs: Vertex) -> Bool {
    return !(lhs == rhs)
}